### ISSUE TITLE
App Crash : Fatal signal 6 (SIGABRT) in /lib/arm/librealmreact.so

### ISSUE THREAD

**Issue by rohitdhall07** [realm/realm-js#1947]
I'm using realm-js 2.2.7 in a react-native app.
The Android build of this app crashes very frequently with below stack trace.

A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 9584 (mqt_js)
D/MotoProvider: uriType: 51 values: updated_at=2018-08-04 19:23:41 display_place=8
D/MotoProvider: display_place: 8
I/SFPerfTracer:      triggers: (rate: 0:0) (0 sw vsyncs) (0 skipped) (3:42625 vsyncs) (5:43319)
I/crash_dump32: obtaining output fd from tombstoned
I//system/bin/tombstoned: received crash request for pid 9533
I/crash_dump32: performing dump of process 9533 (target tid = 9584)
A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
A/DEBUG: Build fingerprint: 'motorola/aljeter/aljeter:8.0.0/OPP27.91-87/89:user/release-keys'
A/DEBUG: Revision: 'p400'
A/DEBUG: ABI: 'arm'
A/DEBUG: pid: 9533, tid: 9584, name: mqt_js  >>> com.girnar.nissan <<<
A/DEBUG: signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
A/DEBUG:     r0 00000000  r1 00002570  r2 00000006  r3 00000008
A/DEBUG:     r4 0000253d  r5 00002570  r6 9306a710  r7 0000010c
A/DEBUG:     r8 7d9f7d80  r9 915515f4  sl 0000006c  fp aef29dc0
A/DEBUG:     ip 00000000  sp 9306a700  lr afe73437  pc afea3aa0  cpsr 200f0010
A/DEBUG:     #00 pc 0004aaa0  /system/lib/libc.so (tgkill+12)
A/DEBUG:     #01 pc 0001a433  /system/lib/libc.so (abort+54)
A/DEBUG:     #02 pc 005197d7  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so (_ZN9__gnu_cxx27__verbose_terminate_handlerEv+226)
A/DEBUG:     #03 pc 004e2689  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so (_ZN10__cxxabiv111__terminateEPFvvE+4)
A/DEBUG:     #04 pc 004e26fd  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so (_ZSt9terminatev+8)
A/DEBUG:     #05 pc 004e2821  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so (__cxa_throw+120)
A/DEBUG:     #06 pc 000c938c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #07 pc 000c940c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #08 pc 00134158  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #09 pc 0016898c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #10 pc 001d4508  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #11 pc 001d963c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #12 pc 001d9728  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #13 pc 001d6b48  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #14 pc 001f498c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #15 pc 0029d1ec  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #16 pc 0023e7c8  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #17 pc 0023f21c  /data/app/com.girnar.nissan-a72YSDR2qR5CQuTP4_5R9Q==/lib/arm/librealmreact.so
A/DEBUG:     #18 pc 000104fd  /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+572)
A/DEBUG:     #19 pc 00010229  /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+32)
A/DEBUG:     #20 pc 000b7a39  /system/lib/libandroid_runtime.so (_ZN7android18NativeMessageQueue8pollOnceEP7_JNIEnvP8_jobjecti+24)
A/DEBUG:     #21 pc 00777585  /system/framework/arm/boot-framework.oat (offset 0x5d3000) (android.app.NativeActivity.onWindowFocusChangedNative [DEDUPED]+92)
A/DEBUG:     #22 pc 000099e5  /dev/ashmem/dalvik-jit-code-cache (deleted)

Crash happens when I logout (executes realm.deleteAll()) and login again (writes again to the DB).

Any help/fix will be appreciated.
Thanks in advance.

**Comment by kneth** [comment]
Do you see it across devices or do you see it for a particular Android device?

Is the database local only or do you sync it to either Realm Cloud or your own Realm Object Server?

**Comment by rohitdhall07** [comment]
I have tested on multiple android devices (Moto G6, Samsung Galaxy S8+, Lenovo device, Moto G2)

Database is local only.

I managed to figured out the exact scenario and fixed the bug at app level.

Scenario:
Realm listener is registered on Realm.Results object. I execute realm.deleteAll() on logout and navigate to login screen.
Now as soon as I login, some data gets inserted in local database. This triggers the listener trying to access the realm.results, but those results are no longer valid causing crash.

In such cases, I have seen crash saying "trying to access data that has been invalidated or deleted".
But this particular case caused crash in shared object library.

**Comment by Sovent** [comment]
I'm probably into similar bug. I've added listener for realm objects and provide callback where I dispatch a redux action with payload `objects.snapshot()` and it crashes my app with similar totally useless stacktrace. For a reason it happens only in a debug mode and works just fine without it.

**Comment by kneth** [comment]
@rohitdhall07 Thanks for the detailed description. Do you call `realm.close()` after `realm.deleteAll()`? Is it possible to call `removeAllListeners()` on your `Realm.Results` object as part of your logout procedure?

@Sovent I suspect it is a different thing. Can you create a separate issue for it (including a stack trace)?

**Comment by rohitdhall07** [comment]
@kneth 
I did not call `realm.close()`.
I do call `removeAllListeners()` on all `Realm.Results` in `componentWillUnmount()` in react-native. But react-navigation (which I have used for navigations) does not unmount the last screen even after I reset the navigation stack with only Login screen in it, after logging out.

I have fixed my issue by adding a check using `isValid()` in `Realm.Results`. It avoids accessing the objects/results if it returns false.
